### ISSUE TITLE
[web] Icon tree shaking doesn't work well on web

### ISSUE THREAD

**Issue by mkustermann** [flutter/flutter#154986]
Let's say we have the following example
```
% flutter create foo && cd foo
```
and modify `lib/main.dart`
```diff
import 'package:flutter/material.dart';

+ final IconData unusedGlobalVariable = 
+     IconData(0xe411, fontFamily: 'MaterialIcons');

void main() {
  runApp(const MyApp());
}
...
```

This added `unusedGlobalVariable` isn't used anywhere. So when compiling the program it should be tree shaken away. Any actual uses of `IconData` are in fact constant uses.

=> Tree shaking of icon fonts should therefore work.

This seems to work as expected when building a native app
```
% flutter build apk --release                          
Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 1384 bytes (99.9% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when building your app.
Running Gradle task 'assembleRelease'...                           46.1s
✓ Built build/app/outputs/flutter-apk/app-release.apk (18.9MB)
```

But when running the web build we get an error:
```
flutter build web --release

This application cannot tree shake icons fonts. It has non-constant instances of IconData at the following locations:
  - file:///.../lib/main.dart:4:5
Target web_release_bundle failed: Error: Avoid non-constant invocations of IconData or try to build again with --no-tree-shake-icons.
#0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
#1      IconTreeShaker._findConstants (package:flutter_tools/src/build_system/targets/icon_tree_shaker.dart:321:7)
<asynchronous suspension>
#2      IconTreeShaker._getIconData (package:flutter_tools/src/build_system/targets/icon_tree_shaker.dart:110:45)
<asynchronous suspension>
#3      IconTreeShaker.subsetFont (package:flutter_tools/src/build_system/targets/icon_tree_shaker.dart:179:5)
<asynchronous suspension>
#4      copyAssets.<anonymous closure> (package:flutter_tools/src/build_system/targets/assets.dart:163:25)
<asynchronous suspension>
#5      Future.wait.<anonymous closure> (dart:async/future.dart:520:21)
<asynchronous suspension>
#6      copyAssets (package:flutter_tools/src/build_system/targets/assets.dart:129:3)
<asynchronous suspension>
#7      WebReleaseBundle.build (package:flutter_tools/src/build_system/targets/web.dart:475:29)
<asynchronous suspension>
...
#20     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:130:9)
<asynchronous suspension>
#21     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
<asynchronous suspension>
#22     main (package:flutter_tools/executable.dart:94:3)
<asynchronous suspension>

Compiling lib/main.dart for the Web...                             30.2s
Error: Failed to compile application for the Web.
```

**Comment by danagbemava-nc** [comment]
Reproducible using the steps outline above. 

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.24.2, on macOS 14.6.1 23G93 darwin-arm64, locale en-US)
    • Flutter version 3.24.2 on channel stable at /Users/deanli/dev/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4cf269e36d (8 days ago), 2024-09-03 14:30:00 -0700
    • Engine revision a6bd3f1de1
    • Dart version 3.5.2
    • DevTools version 2.37.2
```

```
[!] Flutter (Channel master, 3.26.0-1.0.pre.72, on macOS 14.6.1 23G93 darwin-arm64, locale en-US)
    • Flutter version 3.26.0-1.0.pre.72 on channel master at /Users/deanli/dev/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 8a2c3a65fe (9 hours ago), 2024-09-10 22:16:25 -0400
    • Engine revision a26075f9b1
    • Dart version 3.6.0 (build 3.6.0-237.0.dev)
    • DevTools version 2.39.0

```
</details>

**Comment by rakudrama** [comment]
This is a bit of a weird issue.

There is no guarantee that tree-shaking removes specific forms. If tree-shaking does not remove the variable, either because the variable is actually used, or because tree-shaking is insufficiently smart, or because tree-shaking happens after IconTreeShaker is run (compiling for web), then the _program has a bug_ - a non-const reference to a type that should only have const instances.

@mkustermann are you concerned that the tool chain complains about an actual bug, or that, part way through compilation, dart2js has a different intermediate tree-shaking result than the native compiler?

There is work in progress that should improve this. The end result will be something like:

1. Only run the current IconTreeShaker for native programs, where the tool-chain persists a TFA-based simplification of the program.
2. Change the output of IconTreeShaker to a new, more general, common format that _records the use of certain things_.
3. dart2js already has resource reference tracking. Change the output of dart2js to the new common format.
4. Change the Flutter tool to use either the (native) IconTreeShaker output or (web) dart2js output.

This is necessary because the old current format produced by IconTreeShaker is inadequate for web.
dart2js not only does 'tree-shaking', it also partitions the program to allow deferred loading.
Tracking the use of resource references all the way through compilation allows the resources to be deferred-loaded too.
This approach has already been used to partition localized messages so that the initial program download does not contain (or request) messages that cannot possibly be used yet.

The current IconTreeShaker is run before dart2js, too early to guess the partitioning that dart2js will achieve.

I might also be too early for native. The TFA-based simplification removes some definitions, but it does not remove code that becomes dead only after optimizing, e.g. after inlining. For example, it might be that compiled apps contain both Material and Cupertino versions of icons if the predicate used to choose them is too opaque for TFA to resolve.

/cc @mosuem

**Comment by mkustermann** [comment]
> @mkustermann are you concerned that the tool chain complains about an actual bug, or that, part way through compilation, dart2js has a different intermediate tree-shaking result than the native compiler?

This issue is tracking that tree shaking of icon data on the web doesn't (really) work.  It is correct that different backends do different optimizations and may be able to tree shake more or less code. But any non-trivial tree shaker will remove global/static fields that are not referenced (neither read from nor written to). 

> The current IconTreeShaker is run before dart2js

Exactly. So basically tree shaking of icons doesn't work on the web (as it runs before actual dart2js optimiztaions) - even though the `flutter build web` command may indicate it does.

> It might also be too early for native. 

Yes, one may get slightly better results after the C++ AOT compiler. But the 90%+ of things is probably covered by doing it after TFA.

> There is work in progress that should improve this. The end result will be something like:

Wouldn't the idea be that all backends emits the `record_use.json` that @mosuem is working on? dart2wasm&vm would emit it based on visiting kernel AST. dart2js would do it based on it's internal optimized representation.

The IconTreeShaker would then be a 3rd party package that uses `record_use.json` and can tree shake icons based on that.
